apps/condo/domains/notification/tasks.js

```diff
@@ -1,19 +1,98 @@
+const conf = require('@core/config')
 const { createTask } = require('@core/keystone/tasks')
+const { getSchemaCtx } = require('@core/keystone/schema')
+
+const { Message } = require('@condo/domains/notification/utils/serverSchema')
+
 const sms = require('./transports/sms')
 const email = require('./transports/email')
-const { SMS_TRANSPORT, EMAIL_TRANSPORT } = require('./constants')
+const { SMS_TRANSPORT, EMAIL_TRANSPORT, MESSAGE_SENDING_STATUS, MESSAGE_RESENDING_STATUS, MESSAGE_PROCESSING_STATUS, MESSAGE_ERROR_STATUS, MESSAGE_DELIVERED_STATUS } = require('./constants')
+
+const SEND_TO_CONSOLE = conf.NOTIFICATION__SEND_ALL_MESSAGES_TO_CONSOLE || false
 
 const TRANSPORTS = {
     [SMS_TRANSPORT]: sms,
     [EMAIL_TRANSPORT]: email,
 }
 
-async function sendMessageByTransport (messageId, transport) {
-    const adapter = TRANSPORTS[transport]
-    const context = await adapter.prepareMessageToSend(messageId)
-    await adapter.send(context)
+async function _sendMessageByAdapter (transport, adapter, messageContext) {
+    if (SEND_TO_CONSOLE) {
+        console.info(`MESSAGE by ${transport.toUpperCase()} ADAPTER: ${JSON.stringify(messageContext)}`)
+        return [true, { fakeAdapter: true }]
+    }
+    return await adapter.send(messageContext)
+}
+
+async function deliveryMessage (messageId) {
+    const { keystone } = await getSchemaCtx('Message')
+    const messages = await Message.getAll(keystone, { id: messageId })
+    const message = messages[0]
+
+    // TODO(pahaz): extend this logic
+    //  1) we should chose the best transport for the message
+    //  2) then sending it by the transport
+    // For example: if we want to send invite to user and he wants to get receive messages by TG we should schedule
+    // TG transport sending
+    const transport = EMAIL_TRANSPORT
+
+    if (message.id !== messageId) throw new Error('get message by id wrong result')
+    if (message.status !== MESSAGE_SENDING_STATUS && message.status !== MESSAGE_RESENDING_STATUS) {
+        return `already-${message.status}`
+    }
+
+    const baseAttrs = {
+        // TODO(pahaz): it's better to use server side fingerprint?!
+        dv: message.dv,
+        sender: message.sender,
+    }
+
+    const processingMeta = { dv: 1, transport, step: 'init' }
+    await Message.update(keystone, message.id, {
+        ...baseAttrs,
+        status: MESSAGE_PROCESSING_STATUS,
+        deliveredAt: null,
+        processingMeta,
+    })
+
+    try {
+        const adapter = TRANSPORTS[transport]
+        const messageContext = await adapter.prepareMessageToSend(message)
+        processingMeta.step = 'prepared'
+        processingMeta.messageContext = messageContext
+
+        const [isOk, deliveryMetadata] = await _sendMessageByAdapter(transport, adapter, messageContext)
```

Comment:
What's the difference between send/delivery ?

Reply:
Seems like it's _deliverMessageByAdapter

Reply:
And, why we cannot use adapter as pure dependency of deliveryMessage function?

Reply:
delivery is an async task. the message delivery process is handled by the worker. I also add a diagram to README.md